Integrate results of offline discussions into core explainer

tschneidereit opened the pull request:
@nikomatsakis, this should be much closer to what we discussed. I completely reworked the sections dealing with arrays compared to the last PR, and added a section on default values.

tschneidereit committed:
Add links to TOC

tschneidereit committed:
Change struct types to be opaque by default

Fixes #2

tschneidereit committed:
Explain new scheme for array handling
tl;dr: Fixed-sized arrays are now just structs with indexed properties. Additionally, each struct type has an accompanying array constructor and prototype. Explicit ArrayTypes are gone.

Fixes #1

tschneidereit committed:
Changed all vars to consts or lets

`const` for type definitions, `let` for everything else.

tschneidereit committed:
Fill in section on prototypes

Ignores value types for now.

tschneidereit committed:
Primitive type casting clarifications

tschneidereit committed:
Clarify valid contents of structure argument to StructType

tschneidereit committed:
Address @nmatsakis' feedback

Importantly, changes the part on views to make clear only transparent types can be created as views onto preexisting buffers.

The rest is largely bugfixes.

tschneidereit committed:
Clarifications and small fixes to core explainer

tschneidereit committed:
Explain struct type arrays as typed arrays

They're closely related to the existing typed arrays both in intent and in behavior.

tschneidereit committed:
Have assignment throw when source misses fields

tschneidereit committed:
Split `Options` into sub-sections

In preparation of defaults

tschneidereit committed:
Use auto-generated TOC

tschneidereit committed:
Add section on default values to core explainer

tschneidereit committed:
Improve sections ordering of core explainer

And fix up TOC.

No content changes.

tschneidereit committed:
Rewrite of the prototypes section for improved clarity

nikomatsakis commented on core.md:
```diff
@@ -73,11 +92,65 @@ the mechanisms needed to support that.
+
+The second overload is a shortcut for defining struct types with indexed elements
+of a certain type: each entry is an instance of the struct type `elementType`, and
+the length is determined by `length`.
```
It will be important in the spec, at least, to be specific about when each version is chosen -- in particular, it ought to be keyed on the type of the first or second argument, I imagine, so that if people write a dumb forwarder like `foo(args...) { return new StructType(...args); }` (or whatever the splat syntax is) it works for either form.

nikomatsakis commented on core.md:
````diff
@@ -89,10 +162,35 @@ struct would:
+An equivalent definition to this, that'd become unwieldy for large `length`s, would be:
+
+```js
+const PointPairType = new StructType({0: PointType, 1: PointType, length: 2});
````
I don't think we permit things like `length: 2`, so this is not quite _equivalent_, but more "pseudo-code", right?

nikomatsakis replied:
(I presume btw that `length` will be some sort of magic data property, mostly so that it is immutable, though one could also imagine it living on the prototype.)

tschneidereit replied:
I changed this to say that `length` is defined as a property on the type's `prototype`. I actually think this is nicer, and it doesn't require either giving up on this equivalence or introducing singleton types.

nikomatsakis commented on core.md:
````diff
@@ -177,84 +249,105 @@ initialized to 0, fields of type `any` are initialized to `undefined`,
+let line1 = new LineType(example);
 ```
 
 is exactly equivalent to writing:
````
Is this true even with defaults? I think not.

tschneidereit replied:
In this particular example it is, but I clarified that there are caveats to this equivalence (with or without the ability to specify defaults).

nikomatsakis commented on core.md:
```diff
@@ -177,84 +249,105 @@ initialized to 0, fields of type `any` are initialized to `undefined`,
+let coercedPoints = new PointType.array([new PointType(1, 2), new PointType(10, 20)]);
+// Creates an instance as a view onto the given buffer, starting at the given
+// offset and with the given length, both of which are optional.
+let pointsView = new PointType.array(buffer(points), 3, 3);
```
I think this would only be legal if `PointType` were declared as transparent (which, in this example, it is not).

nikomatsakis replied:
Also, assuming `PointType` were transparent, this also implies that distinct views with the same buffer/offset/length are distinct objects (unlike the views below, which are fat pointers). Is that what we actually want? 

In particular, should the following really be `false`?

``` js
new PointType.array(buffer, 3, 3) === new PointType.array(buffer, 3, 3)
```

It seems a bit odd for it to be false, given the way that views and identity work below. In particular, imagine that we have an array of `LineType`:

``` js
const PointType = new StructType({x: f32, y: f32}, {transparent: true});
const LineType = new StructType({start: PointType, end: PointType}, {transparent: true});
const buffer = ...; // some buffer
const lines1 = new LineType(buffer, x, y);
const lines2 = new LineType(buffer, x, y);
lines1 !== lines2; // true
lines1.start === lines2.start; // true (!)
```

Note the last line, which I think is potentially rather surprising. I believe this would follow, though, from the canonicalization rules, since we have two points at the same offset in the same buffer.

So I think we should break compat (intentionally) with typed arrays here and instead go with `PointType.array.view`.

tschneidereit replied:
I changed this to introduce canonicalization between struct arrays. However, I'm somewhat on the fence on this. While I agree with your intuition that the previous behavior would be somewhat surprising, I'm not sure how bad that'd be. The same set of equivalences (and lack thereof) holds for `let a = [{}, {}]; let b = a.slice();`

One could argue that it _would_ also hold for current typed arrays, except it's not observable because they can only contain primitive values.

I think it wouldn't actually matter too much one way or the other, tbh. Except with array canonicalization, we shut the door on making typed object arrays a generalization of the current typed arrays, including respecifying them in terms of typed object arrays. I'm also not sure how crucial this would be, but it'd at least be very nice.

nikomatsakis commented on core.md:
```diff
@@ -274,42 +367,59 @@ before, but with a different offset. The offset is now 16, because it
 which are 8 bytes each).
 
-Unlike arrays and structs, accessing a field of primitive type does
+Unlike for structs, accessing a field of primitive type does
```
this is actually somewhat grammatically ambiguous (as was the previous formulation). Perhaps: "accessing a field of primitive type does not return a typed object, in contrast to fields of struct type."

nikomatsakis commented on core.md:
```diff
@@ -274,42 +367,59 @@ before, but with a different offset. The offset is now 16, because it
+## Assigning fields
 
 When you assign to a field, the backing store is modified accordingly.
 The process is precisely the same as when providing an initial value
```
Not _precisely the same_, due to defaults

nikomatsakis commented:
@tschneidereit seems good, the only _major_ comment I had was https://github.com/nikomatsakis/typed-objects-explainer/pull/21#discussion_r48654708

tschneidereit committed:
Explain mechanism for choosing `StructType` constructor overload

tschneidereit committed:
Fixed alternative definition for second StructType overload

And clarified that the resulting definition has a `length` property on its prototype.

tschneidereit committed:
Clarify caveats to equivalence between initializing and assigning from example object

tschneidereit committed:
Clarify transparency requirements for struct arrays

tschneidereit committed:
Struct arrays are canonicalized

tschneidereit committed:
Fix ambiguity

tschneidereit committed:
Note caveat to initialization/assignment equivalence

tschneidereit commented:
@nikomatsakis, thanks for the feedback. I think I addressed all of it, but see my response on the array canonicalization thing. Something we need to discuss further, I think.

nikomatsakis commented:
@tschneidereit regarding array canonicalization: do you think that this is the only sticking point between typed arrays and typed object arrays? Honestly I don't remember anymore. That does seem like a worthy goal. 

The other reason not to preservere identity is usually optimization, but I'm not sure it applies here. I think it matters much more in the other cases -- and back when arrays were themselves types that could be composed, that mattered more too. But now that arrays are top-level objects, probably not.
